src/test/java/com/uber/cadence/workflow/WorkflowTest.java

```diff
@@ -1832,6 +1833,70 @@ public void testChildWorkflow() {
     assertEquals("HELLO WORLD!", client.execute(taskList));
   }
 
+  private static String childReexecuteId = UUID.randomUUID().toString();
+
+  public interface WorkflowIdReusePolicyParent {
+
+    @WorkflowMethod
+    String execute(WorkflowIdReusePolicy policy);
+  }
+
+  public static class TestChildReexecuteWorkflow implements WorkflowIdReusePolicyParent {
+
+    public TestChildReexecuteWorkflow() {}
+
+    @Override
+    public String execute(WorkflowIdReusePolicy policy) {
+      ChildWorkflowOptions options =
+          new ChildWorkflowOptions.Builder()
+              .setWorkflowId(childReexecuteId)
+              .setWorkflowIdReusePolicy(policy)
+              .build();
+
+      ITestNamedChild child1 = Workflow.newChildWorkflowStub(ITestNamedChild.class, options);
+      Promise<String> r1P = Async.function(child1::execute, "Hello ");
+      String r1 = r1P.get();
+
+      ITestNamedChild child2 = Workflow.newChildWorkflowStub(ITestNamedChild.class, options);
+      String r2 = child2.execute("World!");
+
+      assertEquals(childReexecuteId, Workflow.getWorkflowExecution(child1).get().getWorkflowId());
+      assertEquals(childReexecuteId, Workflow.getWorkflowExecution(child2).get().getWorkflowId());
+      return r1 + r2;
+    }
+  }
+
+  @Test
+  public void testChildAlreadyRunning() {
+    startWorkerFor(TestChildReexecuteWorkflow.class, TestNamedChild.class);
+
+    WorkflowOptions.Builder options = new WorkflowOptions.Builder();
+    options.setExecutionStartToCloseTimeout(Duration.ofSeconds(200));
+    options.setTaskStartToCloseTimeout(Duration.ofSeconds(60));
+    options.setTaskList(taskList);
+    WorkflowIdReusePolicyParent client =
+        workflowClient.newWorkflowStub(WorkflowIdReusePolicyParent.class, options.build());
+    try {
+      client.execute(WorkflowIdReusePolicy.RejectDuplicate);
+      fail("unreachable");
+    } catch (WorkflowFailureException e) {
+      assertTrue(e.getCause() instanceof StartChildWorkflowFailedException);
+    }
+  }
+
+  @Test
+  public void testChildReexecute() {
+    startWorkerFor(TestChildReexecuteWorkflow.class, TestNamedChild.class);
+
+    WorkflowOptions.Builder options = new WorkflowOptions.Builder();
+    options.setExecutionStartToCloseTimeout(Duration.ofSeconds(200));
+    options.setTaskStartToCloseTimeout(Duration.ofSeconds(60));
+    options.setTaskList(taskList);
+    WorkflowIdReusePolicyParent client =
+        workflowClient.newWorkflowStub(WorkflowIdReusePolicyParent.class, options.build());
+    assertEquals("HELLO WORLD!", client.execute(WorkflowIdReusePolicy.AllowDuplicate));
+  }
+
```

Comment:
Could add a test case for starting the second workflow while the first workflow is still open.